### PR TITLE
API summary rework

### PR DESCRIPTION
**Pull request by sschneid**
Bug in #107 squashed; resubmitting for review.

* moves the majority of summary-generation to helper methods
* implements `/summary/:route/:key` dynamic routes, eg.
  * `/summary/tag`
  * `/summary/tag/created_by`
  * `/summary/tag/created_by?from=2015-06-01`
  * `/summary/boot`
  * `/summary/clone`
  * `/summary/clone/duration?from=2015-05-29&to=2015-06-01`
  * ...and so on

### CODE REVIEW

File: lib/vmpooler/api/helpers.rb [hunk 1]
```diff
@@ -158,6 +160,10 @@ def get_tag_metrics(backend, date_str)
             hostname, tag = key.split(':', 2)
           end
 
+          if opts[:only]
+            next unless tag == opts[:only]
```
Comment (colinPL) [inline lib/vmpooler/api/helpers.rb]: Above, `:only` is a boolean (`:only => false`) and here it is being compared to a string (`tag = 'unknown').

Comment (sschneid) [inline lib/vmpooler/api/helpers.rb]: So... it fails!  :)

File: lib/vmpooler/api/v1.rb [hunk 2]
```diff
@@ -121,90 +88,75 @@ def need_token!
...
-        end
+      case params[:route]
+        when 'boot'
+          result = get_task_summary(backend, 'boot', from_date, to_date, :bypool => true, :only => params[:key])
+        when 'clone'
+          result = get_task_summary(backend, 'clone', from_date, to_date, :bypool => true, :only => params[:key])
+        when 'tag'
+          result = get_tag_summary(backend, from_date, to_date, :only => params[:key])
+        else
+          halt 404, JSON.pretty_generate({ 'ok' => false })
```
Comment (colinPL) [inline lib/vmpooler/api/v1.rb]: This should probably be a `400 Bad Request` error. The URL requested does exist, but the parameter is invalid.

Comment (sschneid) [inline lib/vmpooler/api/v1.rb]: It's an invalid (not found) route, though -- eg. `/summary/something` is not found.  This matches the behavior of trying to query a nonexistant VM:
````
$ curl -v --url vmpooler/vm/foo
````
````
< HTTP/1.1 404 Not Found
* Server nginx is not blacklisted
< Server: nginx
< Date: Thu, 04 Jun 2015 20:32:06 GMT
< Content-Type: application/json
< Content-Length: 17
< Connection: keep-alive
< X-Content-Type-Options: nosniff
<
{
  "ok": false
}
````

Comment (colinPL) [inline lib/vmpooler/api/v1.rb]: Makes sense. I thought this was a query param. `/summary?route=something` vs `/summary/something`.

**Comment by colinPL** [comment]
:+1:

**Comment by rji** [comment]
:+1:

**Pull request merged**